### PR TITLE
Various improvements to Tasty Reader [ci: last-only]

### PR DESCRIPTION
**Pull request by bishabosha**
this adds a number of improvements to tasty reader suggested in https://github.com/scala/bug/issues/12197

- block opaque types
- block context function types
- block refinements to `scala.PolyFunction`
- block methods with `@scala.annotation.static`
- block definitions with `@scala.annotation.alpha annotation`
- block generic tuple types
- block large function types
- handle java defined annotations
- block trait parameters for top level traits in a package
- consume values of types that extend traits with parameters

### CODE REVIEW

File: src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala [hunk 1]
```diff
@@ -553,9 +563,11 @@ trait ContextOps { self: TastyUniverse =>
...
         val traitParams = traitParamAccessors(cls).getOrElse(Iterable.empty)
         if (traitParams.nonEmpty) {
-          val parameters = traitParams.map(_.nameString)
-          val msg = s"parameterized trait ${parameters.mkString(s"${cls.nameString}(", ", ", ")")}"
-          unsupportedError(msg)(this.withOwner(cls.owner))
+          log {
+            val parameters = traitParams.map(_.nameString)
+            s"parameterized trait ${parameters.mkString(s"${cls.nameString}(", ", ", ")")}"
+          }
+          cls.updateAttachment(new u.DottyParameterisedTrait(traitParams.toList))
```
Comment (bishabosha) [inline src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala:582]: perhaps theres a more efficient representation we can use for this

**Comment by bishabosha** [comment]
I've rebased and added a few new things in preparation for supporting Scala 3.0.0-M1 release

File: src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala (hunk 1 above)
Comment (bishabosha) [reply src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala:582]: or I guess more lightweight would be that its not necessary to record the actual parameters to present a correct error

**Review by dwijnand — APPROVED** [review]
The changes outside of the DMZ of tasty look fine, so I'm happy for this to land.  CI failed for the rec depth issue that I have a PR for (I restarted the job).

**Comment by bishabosha** [comment]
~I have added another commit for supporting 3.0.0-M1, which should be frozen tastywise now~
~edit as of https://github.com/scala/scala/pull/9271/commits/87062247031543c04e2f5404483ded09277d728d this commit was lost in the history, will need to add it in again~
second edit: it has been added back

**Comment by bishabosha** [comment]
error in CI waiting for #9280

**Comment by bishabosha** [comment]
~I've opened #9286, unless i should rebase and add the changes to this PR~
Edit - closed the PR

**Comment by bishabosha** [comment]
I've now also added the changes to create local field symbols - which means pattern matching on dotty case classes works again

**Review by dwijnand — APPROVED** [review]
LGTM again, thank you Jamie!

**Comment by SethTisue** [comment]
@lrytz you want to give this a once-over, or shall we just merge it?

**Comment by bishabosha** [comment]
I realised I accidentally overwrote the commits that made this work with 3.0.0 - so either this one should wait or I will open a new PR to add the changes again (I need to wait for tomorrow to access my other work laptop)

Edit - second option is probably most simple and then I can add the settings flag to toggle the tasty reader also

Second edit - those commits have been added back to the end of this PR now the CI for 8706224 passed

**Comment by lrytz** [comment]
I'll take a quick look as well.

**Review by dwijnand — APPROVED** [review]
Lovely, thanks for the speedy turnaround.

**Comment by bishabosha** [comment]
I've added back the commits that bring it to date with Scala 3.0.0-M1 nightly

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -1793,6 +1793,16 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
               val ps = psym.info.parents
               if (!ps.isEmpty && !superclazz.isSubClass(ps.head.typeSymbol))
                 pending += ParentSuperSubclassError(parent, superclazz, ps.head.typeSymbol, psym)
+              if (!clazzIsTrait) {
+                // TODO perhaps there can be a flag to skip this when we know there can be no Scala 3 definitions
+                // or otherwise use an optimised representation for trait parameters
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1798]: Use `-Ytasty-reader` https://github.com/scala/scala/pull/9293?

**Review by lrytz — APPROVED** [review]

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (bishabosha) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:1798]: yes that would work nicely

**Pull request merged**